Pool SocketSenders

davidfowl opened the pull request:
- SocketAsyncEventArgs have lots of state on them and as a result are quite big (~350) bytes at runtime. We can pool these since sends are usually very fast and we can reduce the per connection overhead as a result.
- We also allocate one per IOQueue to reduce contention.
- Remove `SocketSenderReceiverBase` and made the `SocketSender` and `SocketReceiver` both derive from SocketAwaitableEventArgs

~Builds on https://github.com/dotnet/aspnetcore/pull/30769~

I ran 2 tests:
- Plaintext - Before allocations 125, after 8 (8 core machine)
- 5000 concurrent websocket connections sending keep alives back and forth - before allocations 5000, after 8


5000 websocket connections over TLS sending keep alives:

SocketReceiver

![image](https://user-images.githubusercontent.com/95136/111026031-f7c5bc00-839c-11eb-8429-0860079c0147.png)

SocketSender

![image](https://user-images.githubusercontent.com/95136/111026062-25126a00-839d-11eb-826f-d2043b36c0e3.png)

davidfowl committed:
Pool SocketSenders
- SocketAsyncEventArgs have lots of state on them and as a result are quite big (~350) bytes at runtime. We can pool these since sends are usually very fast and we can reduce the per connection overhead as a result.
- We also allocate one per IOQueue to reduce contention.
- Fixed buffer list management
- Disposed pool when the transport is disposed
- Added project to slnf so running tests in VS was possible
- Clear the buffer and buffer list before returning to the pool
- This cleans up dumps as the pooled senders don't see references to buffers
while pooled in the queue

JamesNK commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,55 @@
+
+        public void Dispose()
+        {
+            if (!_disposed)
```
Is thread safety an issue when disposing? What happens if there is a sender that hasn't been returned to the queue?

davidfowl replied:
That's not a concern. It'll probably leak. We have that same issue with the memory pool.

davidfowl replied:
To clarify, in the face of ungraceful shutdown, bad things can happen with both the memory pool and now this pool.

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -282,7 +284,10 @@ private async Task ProcessSends()
-                    await _sender.SendAsync(buffer);
+                    _sender = _socketSenderPool.Rent();
+                    await _sender.SendAsync(_socket, buffer);
+                    _socketSenderPool.Return(_sender);
```
If it errors, it's disposed (further up the stack where exceptions are caught). No it shouldn't bein a try/finally, we don't need to  return things to the pool in the error case.

davidfowl commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -282,7 +284,10 @@ private async Task ProcessSends()
+                    _sender = _socketSenderPool.Rent();
+                    await _sender.SendAsync(_socket, buffer);
+                    _socketSenderPool.Return(_sender);
+                    _sender = null;
```
So I can dispose the sender in the error case.

JamesNK replied:
Need some comments explaining why. Is there a test for it?

JamesNK replied:
Comment in the code for future us 😋

davidfowl committed:
Update src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs

Co-authored-by: James Newton-King <james@newtonking.com>

JamesNK commented:
@Tratcher or @jkotalik: Can you take a look? I haven't touched connection-level types much

davidfowl committed:
PR feedback

davidfowl commented:
/azp run

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSender.cs:
```diff
@@ -11,55 +11,56 @@
+            // We clear the buffer and buffer list before we put it back into the pool
+            // it's a small performance hit but it removes the confusion when looking at dumps to see this still
+            // holds onto the buffer when it's back in the pool
+            BufferList = null;
```
Should we preserve the previous code:
```
// The BufferList getter is much less expensive then the setter.
if (_awaitableEventArgs.BufferList != null)
{
    _awaitableEventArgs.BufferList = null;
}
```

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,55 @@
+using System;
```
header

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,55 @@
+{
+    internal class SocketSenderPool : IDisposable
+    {
+        private const int MaxQueueSize = 1024; // REVIEW: Is this good enough?
```
Did you look to see how many were created with 5000 websockets?

davidfowl replied:
yes, 8 on my machine (8 cores).

jkotalik replied:
1024 seems high to me. Is this the maximum number of sends occurring at the same time?

davidfowl replied:
> 1024 seems high to me. Is this the maximum number of sends occurring at the same time?

What number do you like? A multiple of the number of cores?

davidfowl replied:
Also, this is a maximum, it doesn't mean we start with this many or that we'll have this many. It means we won't got past this many.

jkotalik replied:
I understand, but once you get up to 1024, there isn't a way to reduce the amount pooled.

Maybe something like 128 would be better.

jkotalik replied:
> A multiple of the number of cores?

16 * cores?

davidfowl replied:
Right now we allocate one of these queues per IOQueue, we could pick a conservative number like 128. The 1024 was copied from what we used to do in libuv for a similar pool (the WriteQueue) https://github.com/dotnet/aspnetcore/blob/90c45d94ac5f2127063e841964f87188fbda05bc/src/Servers/Kestrel/Transport.Libuv/src/Internal/WriteReqPool.cs#L10.

halter73 replied:
Maybe we should make it 32 so it always fits in a single `ConcurrentQueueSegment`. We'd have to be strict about never going over that in `Return(SocketSender)`.

davidfowl replied:
We don't want to be strict. I'll pick whatever number ya'll want though as long as we don't introduce locking into this code path.

halter73 replied:
You can be strict without locking. You can use `Interlocked.Increment/Decrement`. `ConcurrentQueue.Enqueue` is already doing it's own `Interlocked.Increment` at a minimum.

davidfowl replied:
OK I've complicated the code.

davidfowl replied:
Also, there's no way that I can think of to make this accurate without locking. Review the code and let me know.

halter73 replied:
```suggestion
        // Similar to https://github.com/dotnet/aspnetcore/blob/90c45d94ac5f2127063e841964f87188fbda05bc/src/Servers/Kestrel/Transport.Libuv/src/Internal/WriteReqPool.cs#L10.
        private const int MaxQueueSize = 1024;
```

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,55 @@
+                }
+            }
+        }
+
```
```suggestion
```

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -50,26 +50,30 @@ internal sealed class SocketConnectionListener : IConnectionListener
                         InputOptions = new PipeOptions(_memoryPool, applicationScheduler, transportScheduler, maxReadBufferSize, maxReadBufferSize / 2, useSynchronizationContext: false),
-                        OutputOptions = new PipeOptions(_memoryPool, transportScheduler, applicationScheduler, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false)
+                        OutputOptions = new PipeOptions(_memoryPool, transportScheduler, applicationScheduler, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false),
+                        SocketSenderPool = new SocketSenderPool(awaiterScheduler)
```
Did you try a single pool for all IOQueues?

davidfowl replied:
Nope.

davidfowl replied:
The problem is its tied to the scheduler which is tied to the IOQueue

halter73 replied:
You could make `SocketAwaitableEventArgs._ioScheduler` resettable and make the `PipeScheduler` an argument to `Send/ReceiveAsync` much like `Socket` if you really wanted, but I think having a pool per IOQueue makes sense to reduce contention.

BrennanConroy commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -50,26 +50,30 @@ internal sealed class SocketConnectionListener : IConnectionListener
                         Scheduler = transportScheduler,
                         InputOptions = new PipeOptions(_memoryPool, applicationScheduler, transportScheduler, maxReadBufferSize, maxReadBufferSize / 2, useSynchronizationContext: false),
-                        OutputOptions = new PipeOptions(_memoryPool, transportScheduler, applicationScheduler, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false)
+                        OutputOptions = new PipeOptions(_memoryPool, transportScheduler, applicationScheduler, maxWriteBufferSize, maxWriteBufferSize / 2, useSynchronizationContext: false),
```
Am I crazy, or could PipeOptions be shared?

davidfowl replied:
It can be but I didn't want to write that code 😄 . It can only be shared if the scheduler is inline.

davidfowl committed:
PR feedback

davidfowl commented:
5000 websocket connections over TLS sending keep alives:

SocketReceiver

![image](https://user-images.githubusercontent.com/95136/111026031-f7c5bc00-839c-11eb-8429-0860079c0147.png)

SocketSender

![image](https://user-images.githubusercontent.com/95136/111026062-25126a00-839d-11eb-826f-d2043b36c0e3.png)

jkotalik commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSender.cs:
```diff
@@ -68,18 +72,14 @@ private List<ArraySegment<byte>> GetBufferList(in ReadOnlySequence<byte> buffer)
-            else
-            {
-                // Buffers are pooled, so it's OK to root them until the next multi-buffer write.
-                _bufferList.Clear();
```
How much memory is expected to be used in the bufferlist?

davidfowl replied:
Are you asking how big the list will get?

jkotalik replied:
As we aren't clearing here, how much idle memory will be in the bufferlist?

davidfowl replied:
We clear in Return.

jkotalik commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs:
```diff
@@ -282,7 +284,10 @@ private async Task ProcessSends()
                 if (!buffer.IsEmpty)
                 {
-                    await _sender.SendAsync(buffer);
+                    _sender = _socketSenderPool.Rent();
```
I'm a little confused why we need to rent and return for each send? Can't we rent once at the beginning of ProcessSends and return after (and calling Reset accordingly).

davidfowl replied:
The entire purpose of the change is to avoid that. Today we have one per connection, this reduces it from one per connection to one per send operation, hence why we go from 5000 in the test below to 8.

jkotalik replied:
I see. So this reduces allocations with a pool as well as reducing the amount of time an object is used for.

If we have a limit of 1024, couldn't that be worse though when you get a sudden increase in traffic and then reduces? Is that something to be concerned about here?

davidfowl replied:
Realistically for that to happen, you would need to cause the TCP buffers in the kernel to fill up per connection by sending massive payloads and having the client not read them fast enough. The only way the number of concurrent operations exceeds the number of executing sends is when the sends themselves go async and this is very rare. 

That said, the SocketSender is about 300 bytes a pop and we default to a maximum of 16 IOQueues, if by some glitch we end up with 1024 entries per IOQueue then we'd have 4MB of unfreeable memory.

jkotalik commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public void Return(SocketSender sender)
+        {
+            if (_disposed || _queue.Count > MaxQueueSize)
```
Isn't checking queue.Count > MaxQueueSize thread unsafe? From the way this code looks, we expect multiple calls to Rent and Return at the same time. This code may cause the queue to grow larger that 1024. This may be fine honestly.

davidfowl replied:
Yea but it doesn't matter. There's no hard limit

halter73 replied:
Even though we don't need a hard limit, we should track the count independently for perf/predictability since we query it so frequently.

[ConcurrentQueue.Count](https://github.com/dotnet/runtime/blob/ec5d2eeade1ba4f203b2166c399056e124b4cfb8/src/libraries/System.Private.CoreLib/src/System/Collections/Concurrent/ConcurrentQueue.cs#L246-L349) is pretty harry. It takes a lock and loops if there are more than 2 segments. Based on my reading of the code, there will be more than 2 segments if the pool ever reaches more than 92 items.

davidfowl replied:
But does it really matter?

davidfowl replied:
Are you thinking it'll be a performance problem or a correctness problem?

@stephentoub we really need to expose ConcurrentQueueSegment\<T\> for pooling scenarios like this.

halter73 replied:
My concern is purely perf.

stephentoub replied:
https://github.com/dotnet/runtime/issues/23700

davidfowl replied:
😢 . Do you think it's reasonable to revisit? The concurrent queue with a bounded capacity seemed so simple. I don't think holding off for an object pool (which comes with policy) should overlap.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSender.cs:
```diff
@@ -11,55 +11,59 @@
             }
 
-            _awaitableEventArgs.SetBuffer(MemoryMarshal.AsMemory(memory));
+            SetBuffer(null, 0, 0);
```
```suggestion
            else
            {
                SetBuffer(null, 0, 0);
            }
```

If this does wind up slowing down some benchmarks, the extra calls to SetBuffer and the BufferList setter might be worth looking into.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/SocketConnectionListener.cs:
```diff
@@ -199,6 +206,13 @@ public ValueTask DisposeAsync()
             _memoryPool.Dispose();
+
+            // Dispose any pooled senders
+            foreach (var setting in _settings)
```
We need a better name than `Settings`. Maybe `Context` 😄

davidfowl committed:
PR Feedback
- Keep track of items in the pool separately from the queue count.
- Clear the buffer only if the buffer list is null

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSender.cs:
```diff
@@ -11,55 +11,61 @@
             }
 
-            _awaitableEventArgs.SetBuffer(MemoryMarshal.AsMemory(memory));
+            _bufferList?.Clear();
```
This should probably be put in the `if (BufferList != null)` block. Missed this earlier.

davidfowl replied:
Done

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,76 @@
+
+        public void Return(SocketSender sender)
+        {
+            if (Volatile.Read(ref _disposed))
```
I don't really see the point of using `Volatile.Read` for `_disposed`. There's still a race between `Return` and `Dispose` if they're not synchronized.

davidfowl replied:
Yea, I realized that.

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,76 @@
+            }
+
+            // Add this sender back to the queue if we haven't crossed the max
+            var count = Volatile.Read(ref _count);
```
I would just optimistically `Interlocked.Increment`. If the returned count is greater than `MaxQueueSize`, I'd then `Interlocked.Decrement` and dispose without looping.

davidfowl replied:
Done

halter73 committed:
Update src/Servers/Kestrel/Transport.Sockets/src/Client/SocketConnectionFactory.cs

Co-authored-by: Brennan <brecon@microsoft.com>

davidfowl committed:
More PR feedback

halter73 commented on src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs:
```diff
@@ -0,0 +1,62 @@
+            if (_disposed || Interlocked.Increment(ref _count) > MaxQueueSize)
+            {
+                Interlocked.Decrement(ref _count);
+                // We disposed the queue
```
```suggestion
```

davidfowl committed:
Update src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>
